# Final Project
### Project 2
##### Dan Cardall, Xiaoyan Mi, Pablo

In [50]:
from pricers3 import *
from payoffs import *
from monte_carlo_pricers import *

In [54]:
spot = 100.0
strike = 100.0
rate = 0.06
vol = 0.20
div = 0.03
expiry = 1.0
steps = 52
reps = 10000

In [34]:
beta = -1
the_call = VanillaOption(strike, expiry, call_payoff)
result = naive_monte_carlo_pricer(the_call, spot, rate, vol, div, reps)
result2 =  antithetic_monte_carlo_pricer(the_call, spot, rate, vol, div, reps)
result3 = control_variate_pricer_antithetic(the_call, spot, rate, vol, div, steps, reps, beta)
result4 = control_variate_pricer_antithetic_gamma(the_call, spot, rate, vol, div, steps, reps, beta)

In [35]:
print("Naive Monte Carlo: \n Estimated Price: %8.4f, \n Standard Error: %8.4f" % (result[0], result[1]))
print("Antithetic Monte Carlo: \n Estimated Price: %8.4f, \n Standard Error: %8.4f" % (result2[0], result2[1]))
print("Black-Scholes Delta CV Monte Carlo with Antithetic Sampmling: \n Estimated Price: %8.4f, \n Standard Error: %8.4f" % (result3[0], result3[1]))
print("Black-Scholes Delta & Gamma CV with Antithetic Sampling: \n Estimated Price: %8.4f, \n Standard Error: %8.4f" % (result4[0], result4[1]))


Naive Monte Carlo: 
 Estimated Price:   9.0446, 
 Standard Error:   0.1358
Antithetic Monte Carlo: 
 Estimated Price:   9.1242, 
 Standard Error:   0.1026
Black-Scholes Delta CV Monte Carlo with Antithetic Sampmling: 
 Estimated Price:   9.1129, 
 Standard Error:   0.0107
Black-Scholes Delta & Gamma CV with Antithetic Sampling: 
 Estimated Price:   9.1369, 
 Standard Error:   0.0046


| Method | Estimated Price | Standard Error | Time to Run |
| --- | --- | --- | --- |
| Naive Monte Carlo | 9.0446 | 0.1358 | 997 microseconds |
| Antithetic Monte Carlo | 9.1242 | 0.1026 | 41.1 milliseconds |
| Black-Scholes Delta CV Monte Carlo with Antithetic Sampling | 9.1129 | 0.0107 | 48.4 seconds |
| Black-Scholes Delta & Gamma CV with Antithetic Sampling | 9.1369 | 0.0046 | 1 minute 45 seconds |


### Results
As we can see in the table above, as we move down the different estimation methods the smaller the standard error becomes for our estimated price. The Naive Monte Carlo has the biggest standard error with .1358. When we implemented the Antithetic Monte Carlo, we saw a reduction of .0332 in our standard error. This is because with Antithetic sampling method, we are ulitimately doubling our sample size. 

Next we moved on to the two Black Scholes models. The first one was a Black-Scholes based Delta Control Variate Monte Carlo with Antithetic Sampling method. Here we saw the standard error drop from .1358 in the Naive Monte Carlo to .0107. This was a very large reduction in the standard error. With this method we still get the benefit of artificially doubling our sample size with the antithetic sampling method, but we also introduced a control variate. The control variate acts as a hedge, called the delta hedge.This delta hedge is rebalanced in discrete time intervals which leads to a rebalancing in the holding of the asset.  This process greatly reduces the variance of the payoff of the option.  

Finally, we get to the Black-Scholes Delta and Gamma control variates with Antithetic sampling method. This method has a standard error of .0046. This is a great reduction compared to the .1358 in the Naive Monte Carlo method. Again with this method, we get the benefit of antithetic sampling and delta hedging. Now we add a gamma based control variate which reduces the standard error even further than before. 

Although the Black-Scholes Delta and Gamma control variates with Antithetic sampling method significantly reduced the standard error compared to teh Naive Monte Carlo, it also took a lot longer to compute. The Naive Monte Carlo only took 997 microseconds compared to 1 minute 45 seconds the Black-Scholes Delta and Gamma control variates with Antithetic sampling method took. However, if we wanted to get a standard error close to .0046 with the Naive Monte Carlo, it would take approximately 8,715,331 simulations. The real reason behind this time discrepancy is probably more with our poor coding skills than the actual method, but it would be interesting to further investigate the trade off of standard error with computing time and to try to optimize it. 